### PR TITLE
Introduce ListenOptions.GetServerCertificate

### LINKED ISSUE #28120: Preload default Kestrel config before running Kestrel's configureOptions callback
This issue is inspired by an email from @wiktork. From the email:

> I’m [working] on dotnet-monitor. The tool uses Asp.net Core to host a web server that acts as a control plane for diagnostic tooling. We are currently working on the security work for this tool, in particular supporting https.
>
> I have a scenario where I want to do the following:
>
> - Bind https://localhost:52323
> - Bind http://localhost:52525
>
> - Notice one is https, and the other is http.
>
> - Note the two endpoints serve different information (sensitive data over https, non-sensitive data over http).
> - If there is no default certificate (one generated by dotnet dev-certs) or one specified via Kestrel:Certificates:Default, I would like to successfully bind and listen on the http endpoint, but gracefully fail on the https endpoint.
>
> My initial take on this was to use ConfigureKestrel:
>
> ```c#
> //Simplified code
> ConfigureKestrel((context, options) =>
> {
>         try
>         {
>                //If https
>                options.ListenLocalhost(url.Port, listenOptions => listenOptions.UseHttps() )
>         }
>         catch (InvalidOperationException e)
>         {
>             //This binding failure is typically due to missing default certificate
>             console.Error.WriteLine($"Unable to bind to {url}. Dotnet-monitor functionality will be limited.");
>             console.Error.WriteLine(e.Message);
>         }
> ```
>
> This allows me to selectively UseHttps on an endpoint, but this doesn’t work because the default certificate has not been loaded yet. From my understanding of the code, that happens post configuration:
>
> 1. All configuration occurs (my UseHttps code always throws, indicating there is no default certificate)
> 2. KestrelConfigurationLoader.Load will actually load the default certificate into CertificateManager
> 3. Binding/Running server
>
> At this point I could pass along the certificate myself if it’s present to UseHttps, but I don’t really want to replicate all the default certificate logic from asp.net core. Do you guys have any suggestions on how to accomplish this?

The workaround I provided in the email thread was as follows:

> It's too bad that UseHttps() doesn't automatically use the default certificate from config. UseHttps() will use the installed development certificate if there's one available. I would love to fix it, but this is the unfortunate side-effect of being able change the config section Kestrel reads from in the ConfigureKestrel callback. Fortunately, since you can load the configuration early yourself inside the ConfigureKestrel callback, the fix should be a one-liner. If you add
>
>    `options.Configure(context.Configuration.GetSection("Kestrel")).Load();`
>
> before
>
>    `options.ListenLocalhost(url.Port, listenOptions => listenOptions.UseHttps());`
>
> does that fix your issue?

While I expect this workaround of manually loading config at the start of the ConfigureKestrel callback works, it's not at all obvious that this should be necessary and is an overall bad experience.

If we instead preload default Kestrel config before running configureKestrel's callback should allow the KestrelServerOptions.Listen() and ListenOptions.UseHttps() methods to pick up "Default"/"Development" certificates and "EndpointDefaults" from config. Since ConfigureWebDefaults is responsible for binding the KestrelConfigurationLoader the the "Kestrel" subsection of config [here](https://github.com/dotnet/aspnetcore/blob/b795ac3546eb3e2f47a01a64feb3020794ca33bb/src/DefaultBuilder/src/WebHost.cs#L226), this is also where we should do the preloading.

This is a behavioral breaking change, but maybe we can make this less risky by loading just the defaults, and not the endpoints before ConfigureKestrel's callback is run.

### PR DESCRIPTION
**Pull request by amcasey**
# Introduce ListenOptions.GetServerCertificate

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

An API to retrieve the cert for an endpoint

## Description

Introduce ListenOptions.GetServerCertificate as a way to retrieve the server certificate eagerly (assuming the configuration has been loaded), rather than having to wait until bind-time.

Fixes #28120
Unblocks #45801

### CODE REVIEW

**Comment by amcasey** [comment]
Waiting for API approval (and PR approval, obviously).

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs
Comment (mitchdenny) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:190]: What is the benefit of doing this as an extension method vs. just a regular method?

Comment (mitchdenny) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:190]: Never mind I see its following a pattern here.

**Review by mitchdenny — APPROVED** [review]
LGTM

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:190]: I think we generally err on the side of making it an extension method if it doesn't require access to private state (for encapsulation?).

**Comment by amcasey** [comment]
> Is there an API review for this?

https://github.com/dotnet/aspnetcore/issues/48055

File: src/Servers/Kestrel/Kestrel/test/KestrelConfigurationLoaderTests.cs
Comment (amcasey) [reply src/Servers/Kestrel/Kestrel/test/KestrelConfigurationLoaderTests.cs:412]: Agreed, but I was following the pattern of the nearby tests.

**Review by Tratcher — CHANGES_REQUESTED** [review]
I think this needs a different design. I've left notes on the API review.

**Comment by amcasey** [comment]
We're going to try to fix things without a new API.

**Pull request closed without merging**
